### ISSUE TITLE
async/await strictNullChecks allows run-time bugs 

### ISSUE THREAD

**Issue by mmulet** [Microsoft/TypeScript#16228]
strictNullChecks is not working properly with async functions.
If a type is narrowed before an await statement, it remains narrowed after the await statement.
This can lead to uncaught runtime bugs.
**TypeScript Version:**  2.3.4

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
   
let variable : {member: number} | null
declare function asyncFunc(): Promise<void>
async function main(){
    //early return statement narrows the type
   if(!variable){
       return;
   }
  variable.member; //type is narrowed, so this compiles fine
   await asyncFunc()
   variable.member //Compiler does not have any errors, the type
// is still narrowed after the await statement, but it should not be
}

//This can lead to runtime bugs:

async function RuntimeBug(){
    function timeout(timeout: number): Promise<{}>{
        return new Promise((resolve, reject) => {
            setTimeout(resolve, timeout);
        });
    }
    let nullableVariable : {num: number} | null = {num: 5}
    async function AddAfterASecond(){
        if(!nullableVariable){
            return;
        }
        nullableVariable.num += 1;
        await timeout(1000);
        nullableVariable.num += 1; //runtime error not caught by compiler
    } 

    await Promise.all([
         AddAfterASecond(),
         timeout(500)
         .then(() => {
             nullableVariable = null;
         })
    ])
   //This will throw an error after 1 second 
}

```

**Expected behavior:**

The await statement should undo the type narrowing after an await statement when the variable is declared outside the function scope
```ts
let variable : {member: number} | null
declare function asyncFunc(): Promise<void>
async function main(){
   if(!variable){
       return;
   }
   await asyncFunc()
  //deduced type should be {member: number} | null
   variable.member //should throw a compile time error

}

```
It should work like it does with promises
```ts
let variable : {member: number} | null
declare function asyncFunc(): Promise<void>
function main(){
    if(!variable){
       return;
   }
   asyncFunc()
   .then(() => {
     //deduced type is  {member: number} | null
     variable.member //does throw a compile time error
   })
}

```
**Actual behavior:**
```ts
   //actual behavior
let variable : {member: number} | null
declare function asyncFunc(): Promise<void>
async function main(){
   if(!variable){
       return;
   }
   await asyncFunc()
  //deduced type is  {member: number} 
   variable.member //does not throw a compile time error
}
```

**Comment by andy-ms** [comment]
This happens even without async functions.
```ts
interface I {
    y?: { z: number };
}

const obj: I = { y: { z: 0 } };

function f(x: { y?: { z: number } }) {
    if (x.y !== undefined) {
        doSomethingElse();
        // Thinks 'y' is still defined
        return x.y.z;
    }
}

function doSomethingElse() {
    obj.y = undefined;
}

console.log(f(obj));
```

**Comment by mmulet** [comment]
@andy-ms 
Alright. That makes sense from a trade-off point of view.
As far as I'm concerned, you can close this issue as Won't Fix or Working as Intended if you want.
